src/Channels/Memory.cs

```diff
@@ -0,0 +1,90 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+namespace Channels
+{
+    public struct Memory<T>
+    {
+        private readonly T[] _array;
+        private readonly int _offset;
+        private readonly unsafe void* _memory;
+        private readonly int _memoryLength;
+
+        public unsafe Memory(void* pointer, int offset, int length)
+        {
+            unsafe
+            {
+                _memory = pointer;
+            }
+
+            _array = null;
+            _offset = offset;
+            _memoryLength = length;
+        }
+
+        public unsafe Memory(T[] array, int offset, int length, void* pointer = null)
+        {
+            unsafe
+            {
+                _memory = pointer;
+            }
+
+            _array = array;
+            _offset = offset;
+            _memoryLength = length;
+        }
+
+        public Span<T> Span => this;
+
+        public bool IsEmpty => Length == 0;
+
+        public static implicit operator Span<T>(Memory<T> memory)
+        {
+            if (memory.Length == 0)
+            {
+                return Span<T>.Empty;
+            }
+
+            if (memory._array != null)
+            {
+                return memory._array.Slice(memory._offset, memory.Length);
+            }
+            else
+            {
+                unsafe
+                {
+                    return new Span<T>(memory._memory, memory._memoryLength);
+                }
+            }
+        }
+
+        public unsafe void* UnsafePointer => (byte*)_memory + (Unsafe.SizeOf<T>() * _offset);
```

Comment:
aye, feels like usage should be:

    using(raw = memory.Pin())
    {
        var ptr = (myType*)raw.Pointer; // so annoying that you can't return T*
        // etc
    }

where `Pin` returns a `struct` that knows whether or not it is supposed to unpin. In our case, of pre-pinned memory: this should be almost free. In the case of non-pinned, then `Pin` would obviously have to manually pin the object (not via `fixed`) before fetching the address.

I wonder, however, whether it is simpler and cleaner just to have the more simple:

    public void* SomeName {get;}

which just `throw`s if the data wasn't pre-pinned. Eugh; can't get hugely excited by either!

Side note: should this be done instead with something like the language changes proposed for `Span<T>` ?